src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,384 @@
+			transactionContext.AmbientTransaction.EnlistVolatile(
+				transactionContext,
+				_useConnectionOnSystemTransactionEvents ? EnlistmentOptions.EnlistDuringPrepareRequired : EnlistmentOptions.None);
+			originatingSession.TransactionContext = transactionContext;
+
+			_logger.DebugFormat(
+				"enlisted into system transaction: {0}",
+				transactionContext.AmbientTransaction.IsolationLevel);
+
+			originatingSession.AfterTransactionBegin(null);
+			foreach (var dependentSession in originatingSession.ConnectionManager.DependentSessions)
+			{
+				dependentSession.TransactionContext = new DependentContext(transactionContext);
+				dependentSession.AfterTransactionBegin(null);
+			}
+		}
+
+		public bool IsInActiveSystemTransaction(ISessionImplementor session)
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		public void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				// instead of duplicating the logic, we suppress the system transaction and create
+				// our own transaction instead
+				_adoNetTransactionFactory.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			internal System.Transactions.Transaction AmbientTransaction { get; private set; }
+			public bool ShouldCloseSessionOnSystemTransactionCompleted { get; set; }
+			public bool IsInActiveTransaction { get; internal set; }
+			public bool CanFlushOnSystemTransactionCompleted => _useConnectionOnSystemTransactionEvents;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly bool _useConnectionOnSystemTransactionEvents;
+			private volatile SemaphoreSlim _semaphore;
+			private volatile bool _locked;
+			private readonly AsyncLocal<bool> _bypassWait = new AsyncLocal<bool>();
+
+			public SystemTransactionContext(
+				ISessionImplementor sessionImplementor,
+				System.Transactions.Transaction transaction,
+				bool useConnectionOnSystemTransactionEvents)
+			{
+				_sessionImplementor = sessionImplementor;
+				AmbientTransaction = transaction.Clone();
+				AmbientTransaction.TransactionCompleted += TransactionCompleted;
+				_useConnectionOnSystemTransactionEvents = useConnectionOnSystemTransactionEvents;
+				IsInActiveTransaction = true;
+			}
+
+			public void Wait()
+			{
+				if (_isDisposed)
+					return;
+				if (!_locked && AmbientTransaction.TransactionInformation.Status != TransactionStatus.Active)
+					// Rollback case may end the transaction without a prepare phase, apply the lock.
+					Lock();
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				if (semaphore == null || _bypassWait.Value)
+					return;
+				if (_bypassWait.Value)
+					return;
```

Comment:
`AsyncLocal` used to allow session usage by transaction completion, while other usages are locked.